### PR TITLE
Adjust find next key logic in x/Sync

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
Clean up some of the logic around findNextKey.  Primary changes are reusing the proof to figure out the start key and using a reduced form of the proof for deleted keys.

The findNextKey function is used during sync.  After we have received a proof for some range, we want to determine two things. 
1) was that proof complete for that range or was it limited by the size limit of proofs?
2) If it wasn't a complete proof of the range, what is the next key in the range  that we should query?

findNextKey takes in the largest key proof that was sent along with a range/change proof and compares it with your local data to determine the answer to those two questions.

### CODE REVIEW

File: x/sync/syncmanager.go [hunk 1]
```diff
@@ -542,24 +553,21 @@ func (m *StateSyncManager) setError(err error) {
...
-	// if the last key is equal to the end, then the full range is completed
-	if !bytes.Equal(largestHandledKey, workItem.end) {
-		// find the next key to start querying by comparing the proofs for the last completed key
-		nextStartKey, err := m.findNextKey(ctx, largestHandledKey, workItem.end, proofOfLargestKey)
-		if err != nil {
-			m.setError(err)
-			return
-		}
+func (m *StateSyncManager) completeWorkItem(ctx context.Context, workItem *syncWorkItem, rootID ids.ID, proofOfLargestReceivedKey []merkledb.ProofNode) {
+	largestHandledKey := workItem.end
```
Comment (danlaine) [inline x/sync/syncmanager.go:557]: Can we move this to above `if nextStartKey != nil {` where it's used for the first time

File: x/sync/syncmanager.go [hunk 2]
```diff
@@ -391,84 +391,108 @@ func (m *StateSyncManager) findNextKey(
...
-	if len(rangeEnd) > 0 && bytes.Compare(result, rangeEnd) >= 0 {
+	if len(rangeEnd) > 0 && bytes.Compare(nextKey, rangeEnd) >= 0 {
 		return nil, nil
 	}
 
-	return result, nil
+	return nextKey, nil
+}
+
+func findChildDifference(node1, node2 *merkledb.ProofNode, startByte byte) (byte, bool) {
```
Comment (darioush) [inline x/sync/syncmanager.go:478]: nice!. Can we add a comment for this function and its return value?

File: x/sync/syncmanager.go [hunk 3]
```diff
@@ -391,84 +391,108 @@ func (m *StateSyncManager) findNextKey(
...
+	var (
+		child1, child2 ids.ID
+		ok1, ok2       bool
```
Comment (darioush) [inline x/sync/syncmanager.go:481]: should these be inside the loop?

File: x/sync/syncmanager.go [hunk 4]
```diff
@@ -391,84 +391,108 @@ func (m *StateSyncManager) findNextKey(
...
+		receivedProofNode := receivedProofNodes[receivedProofNodeIndex]
+
+		// deepest node is the proof node with the longest key (deepest in the trie) in the two proofs that hasn't been handled yet.
+		// deepestNodeFromOtherProof is the proof node from the other proof with the same key/depth if it exists, nil otherwise.
+		var deepestNode, deepestNodeFromOtherProof *merkledb.ProofNode
+
+		// select the deepest proof node from the two proofs
+		switch {
+		case receivedProofNode.KeyPath.NibbleLength > localProofNode.KeyPath.NibbleLength:
+			{
```
Comment (darioush) [inline x/sync/syncmanager.go:411]: style nit: I don't think we use these `{}` in case statements anywhere else

File: x/sync/syncmanager.go (hunk 3 above)
Comment (darioush) [reply x/sync/syncmanager.go:481]: nvm, doesn't matter.

File: x/sync/syncmanager.go [hunk 5]
```diff
@@ -361,93 +361,135 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
...
+		if childIndex, hasDifference := findChildDifference(deepestNode, deepestNodeFromOtherProof, startingChildNibble); hasDifference {
+			nextKey = deepestNode.KeyPath.AppendNibble(childIndex).Value
+			break
 		}
 	}
 
-	if result == nil || (len(end) > 0 && bytes.Compare(result, end) >= 0) {
+	// if the result is before or equal to the lastReceivedKey
+	// then use the lastReceivedKey + 0, which is the next largest key after lastReceivedKey
+	if nextKey != nil && bytes.Compare(nextKey, lastReceivedKey) <= 0 {
```
Comment (dboehm-avalabs) [inline x/sync/syncmanager.go:459]: Not sure if this is required, but it isn't an expensive safe guard, so I think it is worth keeping.

File: x/sync/syncmanager.go [hunk 6]
```diff
@@ -361,93 +361,135 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
...
-			return append(start, 0), nil
+		// We only want to look at the children with keys greater than the proofKey.
+		// The proof key has the deepest node's key as a prefix,
+		// so only the next nibble of the proof key needs to be considered.
+		// If the deepest node has the same key as the key being proven,
+		// then all of its children have keys greater than the proof key, so we can start at byte 0
+		startingChildNibble := byte(0)
+
+		// If the deepest node has a key shorter than the key being proven,
+		// so look at the next nibble of the proof key to determine which children have larger keys
```
Comment (darioush) [inline x/sync/syncmanager.go:446]:
```suggestion
		// look at the next nibble of the proof key to determine which children have larger keys
```

**Review by darioush — APPROVED** [review]

File: x/sync/syncmanager.go [hunk 7]
```diff
@@ -350,93 +350,146 @@ func (m *StateSyncManager) getAndApplyRangeProof(ctx context.Context, workItem *
...
+	// If the nextKey is before or equal to the lastReceivedKey
+	// then we couldn't find a better answer than the lastReceivedKey.
+	// Set the nextKey to lastReceivedKey + 0, which is the first key in the open range (lastReceivedKey, rangeEnd)
+	if nextKey != nil && bytes.Compare(nextKey, lastReceivedKey) <= 0 {
+		nextKey = lastReceivedKey
+		nextKey = append(nextKey, 0)
+	}
+
+	// If the nextKey is larger than the end of the range, return nil to signal that there is no next key in range
+	if len(rangeEnd) > 0 && bytes.Compare(nextKey, rangeEnd) >= 0 {
```
Comment (rkuris) [inline x/sync/syncmanager.go:463]: What happens if nextKey is nil here? Seems possible since you check for it at line 457 above.

Comment (dboehm-avalabs) [reply x/sync/syncmanager.go:463]: nil is always smaller than any value, so it would fail the compare.

**Pull request merged**